tiledb/sm/query/readers/dense_reader.cc

```diff
@@ -389,56 +394,121 @@ Status DenseReader::dense_read() {
 
-  auto&& [st, qc_result] = apply_query_condition<DimType, OffType>(
-      subarray,
-      tile_extents,
-      result_tiles,
-      tile_subarrays,
-      tile_offsets,
-      range_info,
-      result_space_tiles,
-      num_range_threads);
-  RETURN_CANCEL_OR_ERROR(st);
-
-  // For `qc_coords_mode` just fill in the coordinates and skip attribute
-  // processing.
-  if (qc_coords_mode_) {
-    fill_dense_coords<DimType>(subarray, qc_result);
-    return Status::Ok();
+  uint64_t t_start = 0;
+  uint64_t t_end = 0;
+  uint64_t subarray_start_cell = 0;
+  uint64_t subarray_end_cell = 0;
+  std::vector<uint8_t> qc_result(
+      condition_.empty() ? 0 : subarray.cell_num(), 1);
+
+  // Keep track of the current var buffer sizes.
+  std::map<std::string, uint64_t> var_buffer_sizes;
+  for (auto& name : names) {
+    var_buffer_sizes.emplace(name, 0);
   }
 
-  // Process attributes.
-  std::vector<std::string> to_read(1);
-  for (auto& buff : buffers_) {
-    auto& name = buff.first;
-    if (name == constants::coords || array_schema_.is_dim(name)) {
-      continue;
+  while (t_end < tile_coords.size()) {
+    stats_->add_counter("internal_loop_num", 1);
+
+    // Get result tiles to process on this iteration.
+    auto&& [ret_t_end, result_tiles] = compute_result_tiles<DimType>(
+        names, condition_names, subarray, t_start, result_space_tiles);
+    t_end = ret_t_end;
+
+    // Add the number of cells to process to subarray_end_cell.
+    for (uint64_t t = t_start; t < t_end; t++) {
+      subarray_end_cell += tile_subarrays[t].cell_num();
     }
 
-    if (condition_.field_names().count(name) == 0) {
-      // Read and unfilter tiles.
-      to_read[0] = name;
-      RETURN_CANCEL_OR_ERROR(
-          read_and_unfilter_attribute_tiles(to_read, result_tiles));
+    // Compute parallelization parameters.
+    uint64_t num_range_threads = 1;
+    const auto num_threads =
+        storage_manager_->compute_tp()->concurrency_level();
+    if ((t_end - t_start) < num_threads) {
+      // Ceil the division between thread_num and tile_num.
+      num_range_threads = 1 + ((num_threads - 1) / (t_end - t_start));
     }
 
-    // Copy attribute data to users buffers.
-    status = copy_attribute<DimType, OffType>(
-        name,
-        tile_extents,
+    // Apply the query condition.
+    auto st = apply_query_condition<DimType, OffType>(
         subarray,
+        t_start,
+        t_end,
+        condition_names,
+        tile_extents,
+        result_tiles,
         tile_subarrays,
         tile_offsets,
         range_info,
         result_space_tiles,
-        *qc_result,
-        num_range_threads);
-    RETURN_CANCEL_OR_ERROR(status);
+        num_range_threads,
+        qc_result);
+    RETURN_CANCEL_OR_ERROR(st);
+
+    // For `qc_coords_mode` just fill in the coordinates and skip attribute
+    // processing.
+    if (qc_coords_mode_) {
+      for (auto& name : names) {
+        clear_tiles(name, result_tiles);
```

Comment:
This seems like we are doing a set of allocations and then immediately throwing away? Ideally we could avoid allocating these at all.

Reply:
We cannot. The unfiltered tiles are required to process the query condition for the attribute/dimensions in names. Once we have the resulting tile bitmaps, we can get rid of the memory as we do here.